lib/crypto-primitives/src/Cryptography/KDF/PBKDF2.hs

```diff
@@ -1,13 +1,140 @@
+-- | key=E11244295150E6713CD76E9A5112347093BDB6ACBF0C8021ABAE29881130B210
+-- | iv =6B7F0C406297F0D90E3BD65AD1FB94BA
+-- | -> generateKey SHA256 10000 (32,16) "password" Nothing
+-- |
+-- | 2. 'key' needed for aes-256-ecb with sha256 no salt
+-- | $ echo -n "password" | openssl enc -aes-256-ecb -pbkdf2 -pass stdin -P -nosalt -iter 10000 -P
+-- | key=E11244295150E6713CD76E9A5112347093BDB6ACBF0C8021ABAE29881130B210
+-- | -> fst $ generateKey SHA256 10000 (32,0) "password" Nothing
+-- |
+-- | 3. 'key' needed for aes-256-ecb with sha512 no salt
+-- | $ echo -n "password" | openssl enc -aes-256-ecb -pbkdf2 -pass stdin -P -md sha512 -nosalt -iter 10000 -P
+-- | key=F4CA507C07D0BD31BC779A08756826A6FD9DD97D43AC25E4B29A0933ABEA03F3
+-- | -> fst $ generateKey SHA512 10000 (32,0) "password" Nothing
+-- |
+-- | 4. 'key' and 'iv' needed for aes-256-cbc with sha256 and salt='00000000' (which is in hex equivalent to '3030303030303030')
+-- | $ echo -n "metadata-secret" | openssl enc -aes-256-cbc -pbkdf2 -pass stdin -P -S 3030303030303030 -iter 10000 -P
+-- | salt=3030303030303030
+-- | key=57FCB522B950BCB78138EECFE7FBE07881E5B49AAA2D1CD761D4495A09A5F16C
+-- | iv =41D88E094C8F202C7DF6654B7F40E5AF
+-- | generateKey SHA512 10000 (32,16) "metadata-secret" (Just "00000000")
+generateKey
+    :: HashAlgorithm h
+    => h
+    -- ^ hash algorithm type, like SHA128, SHA256, SHA512
+    -> Word16
+    -- ^ number of iterations algorithm runs through, more iterations mean slower
+    -> (Word8, Word8)
+    -- ^ key length in bytes, iv length in bytes
+    -> ByteString
+    -- ^ payload
+    -> Maybe ByteString
+    -- ^ salt
+    -> (ByteString, ByteString)
+    -- ^ (key, iv)
+generateKey hashAlgoType iterations (keyLen, ivLen) payload saltM =
+    BS.splitAt (fromIntegral keyLen) whole
+  where
+    whole = KDF.generate
+        (KDF.prfHMAC hashAlgoType)
+        (KDF.Parameters (fromIntegral iterations) (fromIntegral $ keyLen + ivLen))
+        payload
+        (fromMaybe BS.empty saltM)
+
+generateKeyByron
+    :: ByteString
+    -- ^ payload
+    -> ByteString
+generateKeyByron key =
+    fst $ generateKey SHA512 500 (32,0) key (Just salt)
+  where
+    salt = "address-hashing" :: ByteString
+
+generateKeyIcarus
+    :: ByteString
+    -- ^ payload
+    -> ByteString
+generateKeyIcarus key =
+    fst $ generateKey SHA512 2048 (64,0) key (Just salt)
+  where
+    salt = "mnemonic" :: ByteString
+
+-- | Key Derivation Function compliant with https://cips.cardano.org/cips/cip83/
+-- | According to the CIP there are the following parameters for KDF:
+-- |  |------------------------------------------|
+-- |  | cipher	     aes-256-cbc (salted)        |
+-- |  | digest	     pdkdf2                      |
+-- |  | iterations	 10000 (default)             |
+-- |  | key + iv	 32 bytes key + 16 bytes iv  |
+-- |  | salt	     8 bytes                     |
+-- |  |------------------------------------------|
```

Comment:
```suggestion
```

… and this is an ideal candidate for defining a value

```hs
cip83MethodBasic :: PBPKDF2Config SHA256
cip83MethodBasic = PBPKDF2Config
    { hash =  SHA256
    , iterations = 10000
    , …
    , saltLength = 8
    }
```

again, making the code more self-documenting rather than using a comment, which needs to be kept in sync with the code.

However, this value should be defined in a different module. I would suggest a place in the `cardano-wallet` package, for example a module `Cardano.CIP.83`.